index.js

```diff
@@ -37,13 +38,46 @@ function destroy (stream) {
     return stream
   }
 
+  if (stream instanceof zlib.Gzip ||
+      stream instanceof zlib.Gunzip ||
+      stream instanceof zlib.Deflate ||
+      stream instanceof zlib.DeflateRaw ||
+      stream instanceof zlib.Inflate ||
+      stream instanceof zlib.InflateRaw ||
+      stream instanceof zlib.Unzip) {
+    return destroyZlibStream(stream)
+  }
+
   if (typeof stream.destroy === 'function') {
     stream.destroy()
   }
 
   return stream
 }
 
+/**
+ * Destroy a zlib stream
+ *
+ * Zlib streams doesn't have a destroy function in Node.js 6. On top of that
+ * simply calling destroy on a zlib stream in Node.js 8+ will result in a
+ * memory leak. So until that is fixed, we need to call both close AND destroy.
+ *
+ * PR to fix memory leak: https://github.com/nodejs/node/pull/23734
+ *
+ * In Node.js 6+8, it's important that destroy is called before close as the
+ * stream would otherwise emit the error 'zlib binding closed'.
+ *
+ * @param {object} stream
+ * @private
+ */
+
+function destroyZlibStream (stream) {
+  if (typeof stream.destroy === 'function') stream.destroy()
```

Comment:
I looked at a few Node.js versions so far and all the ones that have a .destroy, the .close already calls it. Do both need to be called (for which Node.js versions)? I mainly looked because it kind of felt like this was getting called in the reverse order I would have expected, but since all the .close I've looked at just called .destroy,  I wondered if calling it necessary at all. Just a thought.

Reply:
A destroy function exist on zlib streams in Node.js 0.8 and in Node.js 10, which is pretty inconsistent. So when faces with the task to destroy a zlib stream it can be pretty difficult. I was hoping I could add that functionality to this module. If you feel it's better left to another module, that's ok.

You're right that the order of calling destroy first and then close is a bit weird. And at first I actually also had made it the other way around. But...

[In Node.js 10](https://github.com/nodejs/node/blob/v10.12.0/lib/zlib.js#L425-L428), calling `.close` will also call destroy, but calling `.destroy` will not call `.close` which in turn mean that the handle is not closed either. In nodejs/node#23734 (not yet landed), the new `.destroy` function will also call `.close`

In Node.js [6](https://github.com/nodejs/node/blob/v6.14.4/lib/zlib.js#L479-L482) + [8](https://github.com/nodejs/node/blob/v8.12.0/lib/zlib.js#L355-L358), calling `.close` does not call destroy or vice versa. But in these versions, if you call `.close` prior to calling `.destroy`, an error will be emitted unfortunately. So to make sure it has as few side-effects as possible the order have to be like this at least in Node.js versions prior to v10. In Node.js 10 you could swap the two if you wanted to. But according to nodejs/node#23734 the order shouldn't matter then.

Reply:
Thanks! I was just trying to understand, not asking to really change anything or move it somewhere else. That's why it was just a comment :)